GoogleForCreators/web-stories-wp
# Canvas keys: move to global and make improve focus trapping

## Summary

Changes how canvas focus and keyboard are done.

## Relevant Technical Choices

The key points:
 1. Focus trapper is slightly smarter now: when only one element is selected, it will focus the frame of that element. Otherwise the behavior is unchanged and the canvas itself is selected.
 2. Key handlers have been added to the page carousel for testing and feature completeness. See testing instructions below.
 3. All of the canvas key handlers have been moved to global handlers. Ht #2189.
 4. In most of places if only up/down or left/right were handled, now the relevant components handle both pairs and just no-op them if no implementation is possible. This is done so that key handling is not fragmented: e.g. up/down works and left/right does something else.
 5. Shortcuts are enabled over most of buttons/checkboxes/etc. I.e. clicking Delete on a button is not ignored by Mousetrap. This is a mousetrap feature and it's done by selectively adding "mousetrap" class to the affected DOM elements.
 6. Important! The global shortcuts are by default disallowed from crossing over dialog boundaries. E.g. a global shortcut may or may not work inside a page grid or a color picker. This is configurable. But idea is that shortcuts should not make changes that the user cannot even see, b/c a dialog is blocking their view.
 7. Minor: outlines are added to toggle buttons. This helps testing.

## To-do


## User-facing changes

See relevant bugs reports. Additionally, keyboard shortcuts have been implemented in the page carousel.

## Testing Instructions

See relevant bugs reports.

Additionally, for the page carousel, testing instructions are:
 1. Left/right keys should navigate between the carousel pages.
 2. Command+left/right should reorder the carousel pages.
 3. Delete key should delete the selected carousel page.
 4. Command+D should duplicate the selected page.
 5. Esc key should exit the carousel and go back to the canvas.

---

Fixes #2417
Fixes #1781 
Fixes #1070 
Fixes #1808
